### PR TITLE
Attempt to add Math.atan2 to the inbuilt math function

### PR DESCRIPTION
**Pull request by NigelBreslaw**
[x] Add builtin.
[x] Update docs.
[x] Add test.

### CODE REVIEW

File: internal/compiler/expression_tree.rs [hunk 1]
```diff
@@ -136,10 +137,12 @@ impl BuiltinFunction {
             BuiltinFunction::ACos | BuiltinFunction::ASin | BuiltinFunction::ATan => {
                 Type::Function { return_type: Box::new(Type::Angle), args: vec![Type::Float32] }
             }
-            BuiltinFunction::Log | BuiltinFunction::Pow => Type::Function {
-                return_type: Box::new(Type::Float32),
-                args: vec![Type::Float32, Type::Float32],
-            },
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply internal/compiler/expression_tree.rs]: indeed.

File: internal/compiler/generator/cpp.rs [hunk 2]
```diff
@@ -3351,6 +3351,10 @@ fn compile_builtin_function_call(
             ctx.generator_state.conditional_includes.cmath.set(true);
             format!("std::atan({}) / {}", a.next().unwrap(), pi_180)
         }
+        BuiltinFunction::ATan2 => {
+            ctx.generator_state.conditional_includes.cmath.set(true);
+            format!("std::atan2({}) / {}", a.next().unwrap(), pi_180)
```
Comment (ogoffart) [inline internal/compiler/generator/cpp.rs:3356]:
```suggestion
            format!("std::atan2({}, {}) / {}", a.next().unwrap(), a.next().unwrap(), pi_180)
```

There are two argument :-)

File: internal/compiler/generator/rust.rs [hunk 3]
```diff
@@ -2725,6 +2725,7 @@ fn compile_builtin_function_call(
         BuiltinFunction::ASin => quote!((#(#a)* as f64).asin().to_degrees()),
         BuiltinFunction::ACos => quote!((#(#a)* as f64).acos().to_degrees()),
         BuiltinFunction::ATan => quote!((#(#a)* as f64).atan().to_degrees()),
+        BuiltinFunction::ATan2 => quote!((#(#a)* as f64).atan2().to_degrees()),
```
Comment (ogoffart) [inline internal/compiler/generator/rust.rs:2728]: Two arguments, see log bellow that also has two args.

**Review by ogoffart — COMMENTED** [review]
Nice

**Comment by ogoffart** [comment]
Don't forget to make use of it in a test, such as tests/cases/expr/atan.slint or tests/cases/expr/trigo.slint

File: internal/compiler/expression_tree.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (NigelBreslaw) [reply internal/compiler/expression_tree.rs]: Fixed

File: internal/compiler/lookup.rs [hunk 4]
```diff
@@ -1154,6 +1155,7 @@ impl<'a> LookupObject for NumberExpression<'a> {
             .or_else(|| f("asin", member_function(BuiltinFunction::ASin)))
             .or_else(|| f("acos", member_function(BuiltinFunction::ACos)))
             .or_else(|| f("atan", member_function(BuiltinFunction::ATan)))
+            .or_else(|| f("atan2", member_function(BuiltinFunction::ATan2)))
```
Comment (ogoffart) [inline internal/compiler/lookup.rs:1158]: this one line adds support for something like `x.atan2(y)` in Slint. 
I know that Rust support the same syntax, but i find it a bit weird in that case.
What do you think?

**Review by ogoffart — APPROVED** [review]
looks good.

I just wonder if we really need the postfix notation (`foo.atan2(bar)`) for this function.

Ideally this would work with length as well (so you can do `atan2(34px, 65px)`), but it is a bit more involved (need some extra support in the type system) we can still do that later.

**Review by hunger — APPROVED** [review]

**Comment by NigelBreslaw** [comment]
> worth for a changelog entry ;-) @NigelBreslaw

Done!

**Pull request merged**